File: src/diag/engine.ts
import { Span } from "./span";
import { Code } from "./code";
import { Note } from "./note";

export interface Message {
  code: Code;
  span: Span;
  text: string;
  notes: Note[];
  suggestion?: string;
  context?: string;
  relatedSpans?: { span: Span; text: string }[];
  source?: string;
}

export class Engine {
  private list: Message[] = [];
  private source?: string;

  setSource(source: string): void {
    this.source = source;
  }

  emit(code: Code, span: Span, text: string, context?: string, source?: string): void {
    const msg: Message = { 
      code, 
      span, 
      text, 
      notes: [],
      context,
      source
    };
    this.list.push(msg);
  }

  note(span: Span, text: string, fix?: string): void {
    const msg = this.list[this.list.length - 1];
    if (msg) {
      msg.notes.push({ span, text, fix });
    }
  }

  suggestion(text: string): void {
    const msg = this.list[this.list.length - 1];
    if (msg) {
      msg.suggestion = text;
    }
  }

  related(span: Span, text: string): void {
    const msg = this.list[this.list.length - 1];
    if (msg) {
      if (!msg.relatedSpans) msg.relatedSpans = [];
      msg.relatedSpans.push({ span, text });
    }
  }

  get messages(): Message[] {
    return this.list;
  }

  get failed(): boolean {
    return this.list.length > 0;
  }

  getSource(): string | undefined {
    return this.source;
  }
}

File: src/diag/engine.nim
import std/tables
import ./code
import ./span
import ./note

type
  RelatedSpan* = object
    span*: Span
    text*: string
    file*: string
    source*: string

  Fix* = object
    span*: Span
    text*: string
    title*: string
    file*: string
    source*: string

  Message* = object
    code*: Code
    span*: Span
    text*: string
    notes*: seq[Note]
    suggestion*: string
    context*: string
    relatedSpans*: seq[RelatedSpan]
    fixes*: seq[Fix]
    source*: string
    file*: string

  Engine* = ref object
    list: seq[Message]
    source: string
    file: string
    sources: Table[string, string]

proc newEngine*(): Engine =
  Engine(sources: initTable[string, string]())

proc register*(engine: Engine; file, source: string) =
  engine.sources[file] = source

proc setSource*(engine: Engine; source: string; file: string = "") =
  engine.source = source
  engine.file = file
  if file.len > 0: engine.register(file, source)

proc emit*(engine: Engine; code: Code; span: Span; text: string; context: string = "") =
  var message = Message(code: code, span: span, text: text, context: context)
  if span.file.len > 0 and engine.sources.hasKey(span.file):
    message.source = engine.sources[span.file]
  else:
    message.source = engine.source
  message.file = if span.file.len > 0: span.file else: engine.file
  engine.list.add(message)

proc fix*(engine: Engine; span: Span; text, title: string) =
  if engine.list.len == 0: return
  let file = if span.file.len > 0: span.file else: engine.file
  let source = if file.len > 0 and engine.sources.hasKey(file): engine.sources[file] else: engine.source
  engine.list[^1].fixes.add(Fix(span: span, text: text, title: title, file: file, source: source))

proc note*(engine: Engine; span: Span; text: string; fix: string = "") =
  if engine.list.len > 0: engine.list[^1].notes.add(Note(span: span, text: text, fix: fix))

proc suggestion*(engine: Engine; text: string) =
  if engine.list.len > 0: engine.list[^1].suggestion = text

proc related*(engine: Engine; span: Span; text: string; file: string = ""; source: string = "") =
  if engine.list.len == 0: return
  let resolvedFile = if file.len > 0: file elif span.file.len > 0: span.file else: engine.file
  let resolvedSource = if source.len > 0: source elif resolvedFile.len > 0 and engine.sources.hasKey(resolvedFile): engine.sources[resolvedFile] else: engine.source
  engine.list[^1].relatedSpans.add(RelatedSpan(span: span, text: text, file: resolvedFile, source: resolvedSource))

proc messages*(engine: Engine): seq[Message] = engine.list
proc failed*(engine: Engine): bool = engine.list.len > 0
proc getSource*(engine: Engine): string = engine.source
proc getFile*(engine: Engine): string = engine.file
